add delete home to the storage provider

C0rby opened the pull request:
Added a delete home method to the storage provider to be able to clean up when removing a user.

refs commented on cs3/storage/provider/v1beta1/provider_api.proto:
```diff
@@ -724,3 +726,18 @@ message CreateHomeResponse {
+message DeleteHomeRequest {
+  // OPTIONAL.
+  // Opaque information.
+  cs3.types.v1beta1.Opaque opaque = 1;
```
Having `DeleteHomeRequest` as optional leaks implementation details, for instance, we can call `DeleteHome` call with no arguments and this **COULD** translate into "delete home for the authenticated user", having a side effect.

I'm just saying that we have a chance to make this a pure function and have no side effects by adding a [user](https://cs3org.github.io/cs3apis/#cs3.identity.user.v1beta1.User) to the request 🤷

C0rby replied:
I would be fine for that. The reason I did it like that is that I wanted the API to be similar to CreateHome.

butonic replied:
I think CreateHome should also use a user parameter.

labkode commented:
This API call should be part of an Admin API, we have discussed it in the past and now with this need arising is the right moment to create it. @C0rby can you add this new API here: `cs3/admin/v1beta1/admin_api.proto` with that method?
Note that the method must accept a user ID as argument, this API will be used by site administrators or by robot accounts to automate actions.

C0rby commented:
Cool.
Just for clarification, should I move the method to the admin_api or copy it to have it in both places?

refs commented:
> This API call should be part of an Admin API, we have discussed it in the past and now with this need arising is the right moment to create it. @C0rby can you add this new API here: `cs3/admin/v1beta1/admin_api.proto` with that method?
> Note that the method must accept a user ID as argument, this API will be used by site administrators or by robot accounts to automate actions.

thanks for the clarification, I meant `userid`, not an entire user 👯

labkode commented on cs3/admin/v1beta1/admin_api.proto:
```diff
@@ -0,0 +1,92 @@
+  cs3.types.v1beta1.Opaque opaque = 1;
+  // REQUIRED.
+  // The user for which the action should be performed.
+  cs3.identity.user.v1beta1.UserId user_id = 2;
```
Let's use cs3.identity.user.v1beta1.User instead to give more information to the system when creating an user.

labkode commented on cs3/storage/provider/v1beta1/provider_api.proto:
```diff
@@ -724,3 +729,42 @@ message CreateHomeResponse {
+  cs3.identity.user.v1beta1.UserId owner = 2;
+  // REQUIRED.
+  // The quota for the user.
+  uint64 quota = 3;
```
let's redefine this to:
`uint64 quota_max_bytes`
and add also this field
`uint64 quota_max_files`.

The latter is to limit the number of files a user can have, EOS supports that.

labkode commented on cs3/storage/provider/v1beta1/provider_api.proto:
```diff
@@ -724,3 +729,42 @@ message CreateHomeResponse {
+  cs3.types.v1beta1.Opaque opaque = 1;
+  // REQUIRED.
+  // The owner of the storage.
+  cs3.identity.user.v1beta1.UserId owner = 2;
```
Let's use a full user object here, like in the Admin API, like that the Storage can have more information (like uid/gid) already provided to provision it without having to call the user service.

This function is called only when provisioning users,  the rate of this call will be very low so we can afford to prefetch this info beforehand.

labkode reviewed (changes requested):
@C0rby I've added some comments, overall looks pretty good.

labkode commented:
@C0rby @refs any update on this? Will be nice to merge it in.

C0rby commented:
I rebased the branch.
From my POV we can merge, right @butonic?

kulmann commented:
@C0rby could you rebase please? Would be nice to get this in. :-)

phil-davis commented:
@C0rby are you online this week? Can you rebase this and we can try to push it along. There are things that depend on this that are waiting.

phil-davis commented:
@labkode @butonic @refs looks like time to review.

There is no response from drone CI - but that will be related to the drone troubles over Christmas...

C0rby commented:
Well, new year new luck.

Can we merge this? @refs, @butonic, @labkode

refs commented:
@C0rby it LGTM, could we get @ishank011 to have another pair of eyes? :)

ishank011 commented:
@C0rby I think I'm missing something. Why do we have duplicate definitions for (de)provisioning? IMO, the admin API should serve as a gateway to the RPCs in the storage provider, like the gateway API (but we shouldn't add it there so creating a new one makes sense)

ishank011 commented:
@C0rby looks good but how will `ProvisionStorage` be called? I'm guessing from the `ProvisionUser` method in the admin API? How will we supply the quota parameters to it then?

C0rby commented:
I see your point.
@butonic, what do you say? Add the quota to the ProvisionUser api?

butonic commented:
As discussed at the CS3 we should pick a more generic name for `CreateHome`. @ishank011 proposed `CreateStorageSpace` which captures the difference between creating a storage vs creating something inside a storage. _Something_ being called a space.

Other options were _CreateTree_, _CreateSpace_, _CreateSubTree_, and _CreateRoot_ but they are either too abstract / graph theory influenced (anything with tree or root) or would invent a new term (space).

_Storage space_ seems to be a good middle ground. Furthermore it can be used not only for user homes but also other _spaces_ eg. project or group _spaces_.

The method needs an owner (which can be a user or a group), optionally quota and opaque data as usual.

cc @ishank011 @C0rby

C0rby commented:
@jfd, wouldn't CreateStorageSpace also  need a `path` parameter or something like that?

butonic commented:
CreateStorageSpace where? 

In the gateway? there it would need an indication of which storage should create the space. might be done by path or by storageid.

In a storage provider? No, It is up to the storage provider to determine where to put the new root of the storeage.

In the storage registry? Yes, it needs to know where to mount a storage space. Not only a strorage provitder. But this is a bigger change.

For now, I would limit this change to the StorageProvider Api in the context of the UserProvisioning api: the Gateway can determine the responsible storage provider by calling `storageregistry->GetHome()` and then using a `storageregistry->GetProvider(<home>)`. It can then send the `CreateStorageSpace(<owner>, <quota>)` to it, just as it is currently sending a `CreateHome()`

butonic commented:
@C0rby  While you are at it could you also add an `UpdateStorageSpace` so we can adjust the quota? We need that for the quota related tests.

butonic commented:
blocks https://github.com/owncloud/ocis/issues/1290
blocks https://github.com/owncloud/ocis/issues/1313
https://github.com/owncloud/product/issues/247

C0rby commented:
@ishank011 @labkode, @butonic and I now implemented the storage API how we think it should be.
I removed the admin API again and only added the storage space stuff.

phil-davis commented:
@C0rby is there a PR to `cs3org/reva` that implements this?

C0rby commented:
@phil-davis, not yet. We still have to do that. It will most likely happen within the next sprint.
